portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,28 +14,109 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.FetchOptions.Builder;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import com.google.gson.Gson;
+import com.google.sps.data.Comment;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.util.ArrayList; 
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
+// Servlet responsible for handling comments data 
+@WebServlet("/comment")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    ArrayList<String> messages = new ArrayList<String>();
-    messages.add("Hello");
-    messages.add("Testing JSON");
-    messages.add("I hope this works!");
+    String sort = request.getParameter("sort");
+    String imgRequest = request.getParameter("img");
+
+    // Sort comments according to request parameter
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    if (sort != null && sort.equals("old")) {
+      query = new Query("Comment").addSort("timestamp", SortDirection.ASCENDING);
+    }
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    int numComments = results.countEntities(FetchOptions.Builder.withDefaults());
+
+    // Set the number of comments to print (max)
+    String quantity = request.getParameter("quantity");
+    int max = -1;
+    if (quantity != null && quantity.length() > 0) {
+      max = Integer.parseInt(quantity);
+    } else {
+      max = numComments;
+    }
+
+    // Create Comment objects and add to comments ArrayList
+    ArrayList<Comment> comments = new ArrayList<Comment>();
+    int count = 0;
+    for (Entity entity : results.asIterable()) {
+      long id = entity.getKey().getId();  
+      String name = (String) entity.getProperty("name");
+      String comment = (String) entity.getProperty("comment");
+      long timestamp = (long) entity.getProperty("timestamp");
+      String img = (String) entity.getProperty("img");
+
+      // Only add the comment if it's linked with the current img
+      if (img.equals(imgRequest)) {
+        Comment newComment = new Comment(id, name, comment, timestamp, img);
+        count++;
+        if (count <= max) {
+            comments.add(newComment);
+        }
```

Comment:
You can also do early break at the beginning of the loop.

Reply:
The problem is that if the user wants to display 2 out of 5 comments, only 2 comments will be added to the comments ArrayList, but count still needs to be the total of 5. If I break early, or use the size of the array, the total number of comments will be incorrect and the user will lose the ability to select to see more than 2 comments, since I set the max of the input box to be this total.

Reply:
Maybe instead of doing this, I can create Entities with different names according to the img each comment is attached to. So, instead of having all my Entities be named "Comment" and each have a "img" property, I can make different Entities like "Comment-strawberries" and "Comment-norway", for example. This way the weird counting issue won't happen since I will just query for the specific "Comment-strawberry" entities. Not sure if that made enough sense, but any thoughts?

Reply:
And I would also have to slightly modify my DeleteServlet

Reply:
Hmm can you add comments about "count" so that I can better understand the logic here?

Reply:
I think im unsure about this: "If I break early, or use the size of the array, the total number of comments will be incorrect and the user will lose the ability to select to see more than 2 comments, since I set the max of the input box to be this total."

I think having separate entities for each comment is a good idea. Is this a lot of work?

Reply:
Oh I see, "count" is the upper bound of the number users select to indicate how many comments they want to see. 
Current solution works, so I'll approve and we can optimize it on the next pull request :)

Reply:
Sorry Vikki I just saw your comment above Di's now I think I must have only read the last comment in the thread earlier! I fixed all this logic in the other pull request